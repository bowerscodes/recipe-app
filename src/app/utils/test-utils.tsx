import { ReactNode } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { HeroUIProvider } from "@heroui/react";
import RecipeProvider from "@/context/RecipeContext";

const AllProviders = ({ children }: { children: ReactNode }) => (
  <HeroUIProvider>
    <RecipeProvider>{children}</RecipeProvider>
  </HeroUIProvider>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
