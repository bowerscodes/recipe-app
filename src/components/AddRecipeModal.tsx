"use client";

import { useState } from "react";
import {
  Modal,
  ModalContent,
  ModalHeader,
  ModalBody,
  ModalFooter,
  Button,
  Input,
} from "@heroui/react";

import { useRecipeContext } from "@/context/RecipeContext";
import { Recipe } from "@/types/Recipe";
import { FaPlus, FaTrash } from "react-icons/fa";

interface AddRecipeModalProps {
  isOpen: boolean;
  onOpenChange: (open: boolean) => void;
}

export default function AddRecipeModal({
  isOpen,
  onOpenChange,
}: AddRecipeModalProps) {
  const { addUserRecipe } = useRecipeContext();
  const [formData, setFormData] = useState({
    title: "",
    image: "",
    cookingTime: "",
    ingredients: [""],
    instructions: [""],
  });

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();

    const filteredIngredients = formData.ingredients.filter(
      (ingredient) => ingredient.trim() !== "",
    );
    const filteredInstructions = formData.instructions.filter(
      (instruction) => instruction.trim() !== "",
    );

    if (
      !formData.title ||
      !formData.cookingTime ||
      filteredIngredients.length === 0 ||
      filteredInstructions.length === 0
    ) {
      alert("Please fill in all required fields");
      return;
    }

    const newRecipe: Omit<Recipe, "id"> = {
      title: formData.title,
      image:
        formData.image ||
        "https://community.softr.io/uploads/db9110/original/2X/7/74e6e7e382d0ff5d7773ca9a87e6f6f8817a68a6.jpeg",
      cookingTime: formData.cookingTime,
      ingredients: filteredIngredients,
      instructions: filteredInstructions,
      reviews: [],
      author: "me",
    };

    addUserRecipe(newRecipe);

    // Reset form
    setFormData({
      title: "",
      image: "",
      cookingTime: "",
      ingredients: [""],
      instructions: [""],
    });

    onOpenChange(false);
  };

  const addIngredient = (index: number) => {
    setFormData((prev) => ({
      ...prev,
      ingredients: [...prev.ingredients, ""],
    }));
  };

  const removeIngredient = (index: number) => {
    if (formData.ingredients.length > 1) {
      setFormData((prev) => ({
        ...prev,
        ingredients: prev.ingredients.filter((_, i) => i !== index),
      }));
    }
  };

  const updateIngredient = (index: number, value: string) => {
    setFormData((prev) => ({
      ...prev,
      ingredients: prev.ingredients.map((ingredient, i) =>
        i === index ? value : ingredient,
      ),
    }));
  };

  const addInstruction = () => {
    setFormData((prev) => ({
      ...prev,
      instructions: [...prev.instructions, ""],
    }));
  };

  const removeInstruction = (index: number) => {
    if (formData.instructions.length > 1) {
      setFormData((prev) => ({
        ...prev,
        instructions: prev.instructions.filter((_, i) => i !== index),
      }));
    }
  };

  const updateInstruction = (index: number, value: string) => {
    setFormData((prev) => ({
      ...prev,
      instructions: prev.instructions.map((instruction, i) =>
        i === index ? value : instruction,
      ),
    }));
  };

  const handleInputChange = (field: string, value: string) => {
    setFormData((prev) => ({ ...prev, [field]: value }));
  };

  return (
    <Modal
      isOpen={isOpen}
      onOpenChange={onOpenChange}
      placement="top-center"
      size="2xl"
      scrollBehavior="inside"
    >
      <ModalContent>
        {(onClose) => (
          <form onSubmit={handleSubmit}>
            <ModalHeader>Add new recipe</ModalHeader>
            <ModalBody>
              <Input
                autoFocus
                label="Recipe title"
                placeholder="Enter recipe title"
                variant="bordered"
                value={formData.title}
                onChange={(e) => handleInputChange("title", e.target.value)}
                isRequired
              />
              <Input
                label="Image URL"
                placeholder="Enter image URL (optional)"
                variant="bordered"
                value={formData.image}
                onChange={(e) => handleInputChange("image", e.target.value)}
              />
              <Input
                label="Cooking Time"
                placeholder="e.g., 30 minutes"
                variant="bordered"
                value={formData.cookingTime}
                onChange={(e) =>
                  handleInputChange("cookingTime", e.target.value)
                }
                isRequired
              />

              {/* Ingredients */}
              <div>
                <div className="flex justify-between items-center mb-2">
                  <label className="text-sm font-medium">Ingredients</label>
                  <Button
                    size="sm"
                    color="primary"
                    variant="flat"
                    startContent={<FaPlus />}
                    onPress={() => addIngredient(formData.ingredients.length)}
                  >
                    Add ingredient
                  </Button>
                </div>
                <div>
                  {formData.ingredients.map((ingredient, index) => (
                    <div key={index} className="flex gap-2 mt-2">
                      <Input
                        placeholder={`Ingredient ${index + 1}`}
                        variant="bordered"
                        value={ingredient}
                        onChange={(e) =>
                          updateIngredient(index, e.target.value)
                        }
                        className="flex-1"
                      />
                      <Button
                        size="sm"
                        color="danger"
                        variant="flat"
                        isIconOnly
                        onPress={() => removeIngredient(index)}
                        isDisabled={formData.ingredients.length === 1}
                      >
                        <FaTrash />
                      </Button>
                    </div>
                  ))}
                </div>
              </div>

              {/* Insructions */}
              <div>
                <div className="flex justify-between items-center mb-2">
                  <label className="text-sm font-medium">Instructions</label>
                  <Button
                    size="sm"
                    color="primary"
                    variant="flat"
                    startContent={<FaPlus />}
                    onPress={addInstruction}
                  >
                    Add Step
                  </Button>
                </div>
                <div className="space-y-2">
                  {formData.instructions.map((instruction, index) => (
                    <div key={index} className="flex gap-2">
                      <span className="text-sm text-gray-500 pt-3 w-8">
                        {index + 1}.
                      </span>
                      <Input
                        placeholder={`Step ${index + 1}`}
                        variant="bordered"
                        value={instruction}
                        onChange={(e) =>
                          updateInstruction(index, e.target.value)
                        }
                        className="flex-1"
                      />
                      <Button
                        size="sm"
                        color="danger"
                        variant="flat"
                        isIconOnly
                        onPress={() => removeInstruction(index)}
                        isDisabled={formData.instructions.length === 1}
                      >
                        <FaTrash />
                      </Button>
                    </div>
                  ))}
                </div>
              </div>
            </ModalBody>
            <ModalFooter>
              <Button color="danger" variant="flat" onPress={onClose}>
                Cancel
              </Button>
              <Button color="primary" type="submit">
                Add Recipe
              </Button>
            </ModalFooter>
          </form>
        )}
      </ModalContent>
    </Modal>
  );
}
