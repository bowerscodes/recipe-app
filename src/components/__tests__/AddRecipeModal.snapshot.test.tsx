import { render } from "@/app/utils/test-utils";
import AddRecipeModal from "../AddRecipeModal";

describe("AddRecipeModal Snapshots", () => {
  test("renders AddRecipeModal when closed", () => {
    const mockOnOpenChange = jest.fn();
    const { container } = render(
      <AddRecipeModal isOpen={false} onOpenChange={mockOnOpenChange} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders AddRecipeModal when open", () => {
    const mockOnOpenChange = jest.fn();
    const { container } = render(
      <AddRecipeModal isOpen={true} onOpenChange={mockOnOpenChange} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
