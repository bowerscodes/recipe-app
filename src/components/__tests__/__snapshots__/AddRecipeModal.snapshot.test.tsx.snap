// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AddRecipeModal Snapshots renders AddRecipeModal when closed 1`] = `<div />`;

exports[`AddRecipeModal Snapshots renders AddRecipeModal when open 1`] = `
<div>
  <div
    data-testid="modal"
  >
    <div
      data-testid="modal-content"
    >
      <form>
        <div
          data-testid="modal-header"
        >
          Add new recipe
        </div>
        <div
          data-testid="modal-body"
        >
          <input
            label="Recipe title"
            placeholder="Enter recipe title"
            value=""
            variant="bordered"
          />
          <input
            label="Image URL"
            placeholder="Enter image URL (optional)"
            value=""
            variant="bordered"
          />
          <input
            label="Cooking Time"
            placeholder="e.g., 30 minutes"
            value=""
            variant="bordered"
          />
          <div>
            <div
              class="flex justify-between items-center mb-2"
            >
              <label
                class="text-sm font-medium"
              >
                Ingredients
              </label>
              <button
                color="primary"
                startcontent="[object Object]"
                variant="flat"
              >
                Add ingredient
              </button>
            </div>
            <div>
              <div
                class="flex gap-2 mt-2"
              >
                <input
                  class="flex-1"
                  placeholder="Ingredient 1"
                  value=""
                  variant="bordered"
                />
                <button
                  color="danger"
                  variant="flat"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
          <div>
            <div
              class="flex justify-between items-center mb-2"
            >
              <label
                class="text-sm font-medium"
              >
                Instructions
              </label>
              <button
                color="primary"
                startcontent="[object Object]"
                variant="flat"
              >
                Add Step
              </button>
            </div>
            <div
              class="space-y-2"
            >
              <div
                class="flex gap-2"
              >
                <span
                  class="text-sm text-gray-500 pt-3 w-8"
                >
                  1
                  .
                </span>
                <input
                  class="flex-1"
                  placeholder="Step 1"
                  value=""
                  variant="bordered"
                />
                <button
                  color="danger"
                  variant="flat"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          data-testid="modal-footer"
        >
          <button
            color="danger"
            variant="flat"
          >
            Cancel
          </button>
          <button
            color="primary"
            type="submit"
          >
            Add Recipe
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
